### ISSUE TITLE
Join should return a string + Let's print top-level expressions

### ISSUE THREAD

**Issue by Birch-san** [iovisor/bpftrace#1390]
`join()` should return a string as an expression.

### Is your feature request related to a problem? Please describe.

Just something I noticed (during https://github.com/iovisor/bpftrace/issues/305) when looking at how every function handled strings.

`join()` has too many responsibilities; it needn't reinvent printing. We can use the upcoming non-map `print()` (https://github.com/iovisor/bpftrace/issues/356) for that.

### Describe the solution you'd like

- join() should not print any more
- join() should return a string
- **top-level expressions should be printed!**

A primary thing that distinguishes `join(arg0, arg1)` from `printf("%s %s\n", arg0, arg1)` is that it reduces typing from a common use-case.

`print(join(arg0, arg1))` will become possible with non-map print(), but it's still a mouthful.  
How about implicitly wrapping all top-level expressions in `print()`? So you could do:

```bash
bpftrace -e 'tracepoint:syscalls:sys_enter_renameat2 {
  # this will get wrapped in an implicit print()
  args->common_pid;
  # this will get wrapped in an implicit print()
  join(str(args->oldname), str(args->newname));
  # this will not be printed
  $x = join("hello", "world");
  # this will get wrapped in an implicit print()
  $x;
}'
```

This makes `join()` more cohesive, without breaking scripts that relied on its printing capabilities. Moreover, it provides a nice shorthand for printing _anything_.

**Comment by Birch-san** [comment]
Some details may need hammering out… we don't have a `let` keyword for declaring variables, so `$x;` may be intended as a variable declaration.

But the other cases I think are more clear-cut.

**Comment by Birch-san** [comment]
```python
$x = join("hello", "world");
```

Not sure whether assignment expressions have values in bpftrace… can you do `if($x = 123)` or `$x = $y = 123;`?

But in this particular case: even if it's a "top-level value", I expect the user doesn't want it printed.

**Comment by rgolangh** [comment]
@Birch-san +1 for `join` should just return the string, and also 
I would love to see `sprintf(char *fmt, ...)` added.

I have a use case that I want a map where I want to compose the key from several strings and variables:
```
{ @fileread[sprintf("readlink -f /proc/%d/fd/%d", uid, args->fd)] = count(); }'
```

**Comment by fbs** [comment]
> ```
> sprintf("readlink -f /proc/%d/fd/%d", uid, args->fd)
> ```

What is the exact usecase here, why won't printf (or system?) work?

**Comment by rgolangh** [comment]
I actually want system() to return the stdout here and not print, because I wanted to resolve 
a file descriptor to the real file (and use is in a map). To return to subject of this issue I think join() and system() should return a string and then chain printf if needed.

**Comment by fbs** [comment]
> I actually want system() to return the stdout here and not print, because I wanted to resolve

I don't think it will be possible to do what you want. `system` happens in userspace, triggered by an event from the kernel (compiled into the bpf prog). The BPF prog cannot execute random system binaries, that would mean your  probe will be blocked until the binary completes, if it ever does (the probe might block the binary). We could detect the construct and do the map writing in user space, but that will cause race conditions (as it might take hours for userspace to get the result).

Sprintf should be possible but will need kernel support,

**Comment by Birch-san** [comment]
Looks like this is a duplicate of https://github.com/iovisor/bpftrace/issues/26

**Comment by jordalgo** [comment]
Closing in favor of https://github.com/bpftrace/bpftrace/issues/26
